cpp/benchmarks/common/generate_input.cu

```diff
@@ -274,128 +371,127 @@ struct stored_as<T, std::enable_if_t<std::is_same_v<T, bool>>> {
-                     data,
-                     null_mask,
-                     row);
-    }
+  rmm::device_uvector<T> data(0, rmm::cuda_stream_default);
+  rmm::device_uvector<bool> null_mask(0, rmm::cuda_stream_default);
 
-    if (avg_run_len > 1) {
-      int const run_len = std::min<int>(num_rows - row, std::round(run_len_dist(engine)));
-      for (int offset = 1; offset < run_len; ++offset) {
-        set_element_at(
-          data[row], cudf::bit_is_set(null_mask.data(), row), data, null_mask, row + offset);
-      }
-      row += std::max(run_len - 1, 0);
-    }
+  if (cardinality == 0) {
+    data      = value_dist(engine, num_rows);
+    null_mask = valid_dist(engine, num_rows);
+  } else {
+    // generate n samples. and gather.
+    auto sample_indices =
+      sample_indices_with_run_length(avg_run_len, cardinality, num_rows, engine);
+    data      = rmm::device_uvector<T>(num_rows, rmm::cuda_stream_default);
+    null_mask = rmm::device_uvector<bool>(num_rows, rmm::cuda_stream_default);
+    thrust::gather(
+      thrust::device, sample_indices.begin(), sample_indices.end(), samples.begin(), data.begin());
+    thrust::gather(thrust::device,
+                   sample_indices.begin(),
+                   sample_indices.end(),
+                   samples_null_mask.begin(),
+                   null_mask.begin());
   }
 
-  // cudf expects the null mask buffer to be padded up to 64 bytes. so allocate the proper size and
-  // copy what we have.
-  rmm::device_buffer result_bitmask{cudf::bitmask_allocation_size_bytes(num_rows),
-                                    rmm::cuda_stream_default};
-  cudaMemcpyAsync(result_bitmask.data(),
-                  null_mask.data(),
-                  null_mask.size() * sizeof(cudf::bitmask_type),
-                  cudaMemcpyHostToDevice,
-                  rmm::cuda_stream_default);
+  auto [result_bitmask, null_count] =
+    cudf::detail::valid_if(null_mask.begin(), null_mask.end(), thrust::identity<bool>{});
 
   return std::make_unique<cudf::column>(
     cudf::data_type{cudf::type_to_id<T>()},
     num_rows,
-    rmm::device_buffer(data.data(), num_rows * sizeof(stored_Type), rmm::cuda_stream_default),
-    std::move(result_bitmask));
+    data.release(),
+    profile.get_null_frequency() < 0 ? rmm::device_buffer{} : std::move(result_bitmask));
 }
 
-/**
- * @brief Class that holds string column data in host memory.
- */
-struct string_column_data {
-  std::vector<char> chars;
-  std::vector<cudf::size_type> offsets;
-  std::vector<cudf::bitmask_type> null_mask;
-  explicit string_column_data(cudf::size_type rows, cudf::size_type size)
+struct valid_or_zero {
+  template <typename T>
+  __device__ T operator()(thrust::tuple<T, bool> len_valid) const
   {
-    offsets.reserve(rows + 1);
-    offsets.push_back(0);
-    chars.reserve(size);
-    null_mask.insert(null_mask.end(), null_mask_size(rows), ~0);
+    return thrust::get<1>(len_valid) ? thrust::get<0>(len_valid) : T{0};
```

Comment:
Can this be rewritten to use `thrust::pair` instead so that you can just use `.first` and `.second`?

Reply:
Also, I think the type should automatically be constructed correctly.
```suggestion
    return thrust::get<1>(len_valid) ? thrust::get<0>(len_valid) : {0};
```

Reply:
since this is used with `thrust::make_zip_iterator`, `tuple` is required.
Also, `{0}` throws error `Initializer list cannot be used on the right hand side of operator ':'clang(init_list_bin_op)`